task/control-plane-agent/src/inventory.rs

```diff
@@ -0,0 +1,201 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use core::fmt::{self, Write};
+use gateway_messages::{
+    sp_impl::DeviceDescription, DeviceCapabilities, DevicePresence, SpComponent,
+};
+use task_validate_api::DEVICES as VALIDATE_DEVICES;
+use task_validate_api::{Validate, ValidateError, ValidateOk};
+use userlib::UnwrapLite;
+
+// Most of the devices we report to MGS come from asking the `validate` task,
+// but we have a handful that we describe ourself; they're listed by
+// `OurInventory`, and we (in `Inventory` below) logically glue together
+// `OurInventory`'s devices with `VALIDATE_DEVICES`.
+mod our_inventory;
+
+use our_inventory::OurInventory;
+
+userlib::task_slot!(VALIDATE, validate);
+
+pub(crate) struct Inventory {
+    validate_task: Validate,
+    ours: OurInventory,
+}
+
+impl Inventory {
+    pub(crate) fn new() -> Self {
+        let () = ASSERT_EACH_DEVICE_FITS_IN_ONE_PACKET;
+
+        Self {
+            validate_task: Validate::from(VALIDATE.get_task_id()),
+            ours: OurInventory::new(),
+        }
+    }
+
+    pub(crate) fn num_devices(&self) -> usize {
+        self.ours.num_devices() + VALIDATE_DEVICES.len()
+    }
+
+    pub(crate) fn device_description(
+        &self,
+        index: usize,
+    ) -> DeviceDescription<'static> {
+        // If `index` is in `0..self.ours.num_devices()`, defer to it;
+        // otherwise, subtract `self.ours.num_devices()` to shift it into the
+        // range `0..VALIDATE_DEVICES.len()` and ask `validate`.
+        let index = match index.checked_sub(self.ours.num_devices()) {
+            // Subtraction failure; `index` is in `0..self.ours.num_devices()`.
+            None => return self.ours.device_description(index),
+            // Subtraction success; `index` has been shifted down to be in the
+            // range `0..VALIDATE_DEVICES.len()`.
+            Some(index) => index,
+        };
+
+        let device = &VALIDATE_DEVICES[index];
+
+        let presence = match self.validate_task.validate_i2c(index) {
+            Ok(ValidateOk::Present | ValidateOk::Validated) => {
+                DevicePresence::Present
+            }
+            Ok(ValidateOk::Removed) | Err(ValidateError::NotPresent) => {
+                DevicePresence::NotPresent
+            }
+            Err(ValidateError::BadValidation) => DevicePresence::Failed,
+            Err(ValidateError::Unavailable | ValidateError::DeviceOff) => {
+                DevicePresence::Unavailable
+            }
+            Err(ValidateError::DeviceTimeout) => DevicePresence::Timeout,
+            Err(ValidateError::InvalidDevice | ValidateError::DeviceError) => {
+                DevicePresence::Error
+            }
+        };
+
+        let mut component = FmtComponentId::default();
```

Comment:
Add a comment here explaining why this won't fail (I assume it's checked by `ASSERT_EACH_DEVICE_FITS_IN_ONE_PACKET` above?)

Reply:
Almost; it's checked by the stuff in the `max_num_devices` submodule. Will add a comment to that effect.